### ISSUE TITLE
Support Firefox browser 

### ISSUE THREAD

**Issue by brian-mann** [cypress-io/cypress#1096]
PR: https://github.com/cypress-io/cypress/pull/1359

- [x] Update `packages/launcher` to find Firefox, Firefox Dev Edition, and Firefox Nightly
- [x] Launch Firefox with profile + args (look at FirefoxDriver for reference)
- [x] Convert `packages/extension` to `packages/webextension` via [mozilla polyfill](https://github.com/mozilla/webextension-polyfill)
- [x] Add `server/lib/browsers/firefox.coffee` to implement opening the browser
- [x] Ensure all driver integration tests pass on firefox
- [x] Ensure all e2e tests pass on firefox
- [x] Ensure all CSS / visuals work as intended
- [x] Update our docker images to include firefox https://github.com/cypress-io/cypress-docker-images/issues/27
- [x] Make decisions on running e2e or integration tests on firefox in CI
- [x] Add browser-related property and methods to `Cypress`
- [x] Merge in `before:browser:launch` changes https://github.com/cypress-io/cypress/pull/6293

**Comment by brian-mann** [comment]
First passing testing running Cypress in Firefox...

![screen shot 2018-01-09 at 11 13 11 pm](https://user-images.githubusercontent.com/1268976/34789016-9ae883be-f60a-11e7-9cdd-0416b713f6c8.png)

**Comment by Velveeta** [comment]
I and my organization cannot express our ❤️s for you enough right now... I shall drink to your success today... and I shall drink right now........ and also 4 hours ago...

**Comment by aaronherman** [comment]
Is there any update on this?

**Comment by brian-mann** [comment]
The branch is here: https://github.com/cypress-io/cypress/tree/issue-1096-firefox-support

Chris had to pivot and help with another project but he's actually back to work on Firefox support this week. It'll land whenever all the tests pass. There are some failing tests which are causing us to have to update portions of the driver - some of which take a lot of research to reproduce and understand why they're failing. It's nearly there but a bit more to go...

**Comment by morficus** [comment]
Just a heads up... your Docker image includes instructions on how to run Cypress using Firefox, but there is no Firefox support yet. It may be advisable to either link to this issue in Docker Hub or put something in the Docker Hub repo info about Firefox being installed but not usable yet.

https://hub.docker.com/r/cypress/browsers/

**Comment by ikornienko** [comment]
It seemed to be so close, a lot of commits flying into the corresponding branch, and then it seems like some higher priority stuff came up, no activity since the end of Feb... @brian-mann can you please share an update on it? Is there a timeline change for adding this support or is it postponed indefinitely because of some hard to resolve problem?

**Comment by jennifer-shehane** [comment]
Indeed, we have deemed some other work as higher priority. We are mainly working on releasing [this Epic](https://github.com/cypress-io/cypress/issues/681) which will pave the way for parallelizing specs.

We're sorry for the delay on Firefox support - it is still on our Roadmap and the work that has been committed is still of value to our project.

**Comment by serhiibr** [comment]
@jennifer-shehane , thanks for update. Is there an approximate date when the first test in Firefox browser can be runned?

**Comment by pgroot91** [comment]
@Alon87 My guess is they are trying to push hard to finally push out version 3.0 before they are start focusing and finishing on the rest. Alse red somewhere that they are aiming for a 2 week release cycle after v3. It looks like everything is standing still since the last update was probably 2 months ago. Also can't wait to see support for FF.

**Comment by willischu** [comment]
bump

**Comment by willischu** [comment]
The Cypress team is killing it!

**Comment by michaelenglo** [comment]
Do you already have a set date for when ff support will be shipped?

**Comment by pointblack** [comment]
@brian-mann  looking forward to cypress. Really great job you guys did with cypress. Testing finally is no pain anymore

**Comment by morficus** [comment]
If you guys need beta testers to try out the Firefox stuff, hit me up ✋

**Comment by CaroBernat** [comment]
I'm up for beta testing the Firefox stuff too!

**Comment by atilacamurca** [comment]
<https://github.com/deepsweet/foxr> - :fox_face: Node.js API to control Firefox

Hope that this could help in some way.

**Comment by gaetanzip** [comment]
Hi, any update on FF support ? :)

**Comment by ShivaniGabaKreditech** [comment]
Hi team , 
Is there any update on support for FF? Looking forward, thanks

**Comment by willklein** [comment]
@microtema See #310 for more general cross-browser updates. I'm a little burned out hearing people ask the same question again, on this issue, and on Twitter. I'm not sure any developer enjoys being asked "is it done yet" over and over again. It's a really non-trivial problem to solve. It's probably very draining for their team to hear this AND to have been working really hard for many months on this feature. 😅

I'm also glad to see the shift in priorities:

> Cross browser testing is now a front and center major focus here at Cypress.

As a Cypress user, I recommend others not hold out for cross-browser support. One day it will be ready and you can revisit your tooling of warranted. For my team, we are getting 99% of the value from fast e2e test development and execution. We use Selenium for the small set of UI cases that need cross-browser testing.

**Comment by jennifer-shehane** [comment]
Brian Mann (@brian-mann) gave some updates on the work we've been doing for cross browser support as well as other future features on our recent webinar - so check it out if you have your eye on this feature: https://youtu.be/FfqD1ExUGlw?t=2079

**Comment by Mathyn** [comment]
Any updates on this?

**Comment by mgol** [comment]
Right, there haven't been any new commits on the [issue-1096-firefox-support](https://github.com/cypress-io/cypress/tree/issue-1096-firefox-support) branch since October. It'd be good to hear about progress updates from time to time.

My company may be evaluating a new e2e solution soon and having some knowledge about Firefox support progress would help.

**Comment by metasean** [comment]
It looks like the latest info is available on issue [#3207 - Cross Browser issue tracker](https://github.com/cypress-io/cypress/issues/3207).  

Unfortunately, other than the almost completely unchecked todo list at the top, there doesn't appear to be any meaningful movement on the Firefox front 😟

**Comment by digitarald** [comment]
In case people are watching this issue, active development is happening in #1359

**Comment by mayrbenjamin92** [comment]
We would love to have FF support <3!

**Comment by L0wry** [comment]
Hey pals,
So so so keen to get this over the line. Anything I can do to help?

**Comment by kuntemandar** [comment]
Hello Pals,

has anyone tried the pre-release version for firefox like here: 
https://github.com/cypress-io/cypress/pull/1359#commitcomment-36508049

i am trying to understand what the steps in https://on.cypress.io/installing-cypress#Install-pre-release-version are for? is it for users to try out the features in the unmerged develop branch. If so i has anyone tried it. I installed the binary but not able to see firefox in the list of browsers when i do `npx cypress open`

**Comment by ludakhris** [comment]
@kuntemandar suspect you might be a bit  early lol. They haven’t released the binaries yet, their just working the documentation for it’s eventual release.

**Comment by kuntemandar** [comment]
> @kuntemandar suspect you might be a bit early lol. They haven’t released the binaries yet, their just working the documentation for it’s eventual release.

Got it, thanks for clarifying, saved sometime for me... 😄

**Comment by jennifer-shehane** [comment]
We are still actively working on releasing Firefox support. Currently the release is blocked by a garbage collection issue that causes Firefox to crash during some tests during `cypress run`. We're working with the Firefox team and also trying several creative workarounds to prevent the crashing.

**Comment by Baastieng** [comment]
Thank you so much for your hard work. That's been a while I'm waiting for this Firefox compatibility :) Can't wait to use it 💪

**Comment by uptownhr** [comment]
Does the release of playwright mean any possibilities here for faster browser support?

**Comment by tikolakin** [comment]
Additionally, release of firefox support in puppeter, does it help with getting cypress firefox support released?
https://twitter.com/mathias/status/1222183802870607872

**Comment by Velveeta** [comment]
I certainly hope so, because you all started off as a hot contender, but are falling behind the market to Puppeteer and Playwright at this point...
